File: src/main/clients/openrgb/effects/RainbowWave.ts
import { RGBColor } from '@main/clients/openrgb/client/classes/RGBColor';
import { ZoneType } from '@main/clients/openrgb/client/interfaces/ZoneType';
import { AbstractEffect } from '@main/clients/openrgb/effects/base/AbstractEffect';

class RainbowWave extends AbstractEffect {
  public constructor() {
    super('Rainbow Wave', false);
  }

  protected async applyEffect(): Promise<void> {
    const longestZone = Math.max(
      ...new Set(
        this.devices.map((el) =>
          Math.max(...new Set(el.zones.map((el1) => el1.matrix?.width || el1.ledsCount)))
        )
      )
    );

    const rainbow: Array<number> = Array(longestZone);
    for (let idx = rainbow.length; idx >= 0; idx--) {
      rainbow[idx] = (rainbow.length - idx) * 5;
    }
    for (let i = 0; this.isRunning; i = (i + 1) % longestZone) {
      this.devices.forEach((dev) => {
        const colors = Array(dev.leds.length).fill(new RGBColor(0, 0, 0));
        let offset = 0;
        dev.zones.forEach((zone) => {
          if (zone.type == ZoneType.MATRIX) {
            for (let r = 0; r < zone.matrix!.height; r++) {
              for (let c = 0; c < zone.matrix!.width; c++) {
                const rainbowIndex = Math.floor(rainbow.length * (c / zone.matrix!.width));
                colors[offset + zone.matrix!.keys![r]![c]!] = RGBColor.fromHSV(
                  rainbow[rainbowIndex],
                  1,
                  1
                );
              }
            }
          } else {
            for (let l = 0; l < zone.ledsCount; l++) {
              const rainbowIndex = Math.floor(rainbow.length * (l / zone.ledsCount));
              colors[offset + l] = RGBColor.fromHSV(rainbow[rainbowIndex], 1, 1);
            }
          }
          offset += zone.ledsCount;
        });
        this.setColors(dev, colors);
      });
      await this.sleep(100);

      for (let idx = rainbow.length - 1; idx >= 1; idx--) {
        rainbow[idx] = rainbow[idx - 1];
      }
      rainbow[0] = (rainbow[1] + 5) % 360;
    }
    this.hasFinished = true;
  }
}

export const rainbowWave = new RainbowWave();

File: lib/clients/openrgb/effects/rainbow_wave.py
from .base.abstract_effect import AbstractEffect
from ..sdk.utils import RGBColor, ZoneType
from ....utils.singleton import singleton

import math

@singleton
class RainbowWave(AbstractEffect):
    def __init__(self):
        super().__init__("Rainbow wave")

    def apply_effect(self):
        longest_zone = max(
            set(
                max(set(zone.mat_width or len(zone.leds)
                    for zone in el.zones)) 
                for el in self.devices
            )
        )

        rainbow = [0] * longest_zone
        for idx in range(len(rainbow) - 1, -1, -1):
            rainbow[idx] = (len(rainbow) - idx) * 7

        i = 0
        while self.is_running:
            for dev in self.devices:
                colors = [RGBColor(0, 0, 0)] * len(dev.leds)
                offset = 0
                for zone in dev.zones:
                    if zone.type == ZoneType.MATRIX:
                        for r in range(zone.mat_height):
                            for c in range(zone.mat_width):
                                if zone.matrix_map[r][c] is not None:
                                    rainbow_index = math.floor(len(rainbow) * (c / zone.mat_width))
                                    colors[offset + zone.matrix_map[r][c]] = RGBColor.fromHSV(
                                        rainbow[rainbow_index], 1, 1
                                    )
                    else:
                        for l in range(len(zone.leds)):
                            rainbow_index = math.floor(len(rainbow) * (l / len(zone.leds)))
                            colors[offset + l] = RGBColor.fromHSV(rainbow[rainbow_index], 1, 1)
                    offset += len(zone.leds)
                self.set_colors(dev, colors)

            self.sleep(0.15)

            for idx in range(len(rainbow) - 1, 0, -1):
                rainbow[idx] = rainbow[idx - 1]
            rainbow[0] = (rainbow[1] + 7) % 360

rainbow_wave = RainbowWave()